<a href="https://colab.research.google.com/github/HarryKmK13/Data-Mineing-and-analytics-26/blob/main/Khant_KyawMin_Lab_4__Neural_Networks.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# DMA Fall 2026
Please make a copy of this notebook. Make sure you fill in any place that says `YOUR CODE HERE` or `YOUR ANSWER HERE`, as well as your name below:


In [ ]:
NAME = "Kyaw Min Khant"

# Lab 4: Neural Networks #
**Please read the following instructions very carefully**

## Working on the assignment / FAQs
- **Always use the seed/random_state as *42* wherever applicable** (This is to ensure repeatability in answers, across questions, students and coding environments).
- All questions will be graded manually.
- Most questions have two cells:
  - A code cell for your work/code
  - A text cell for giving your final answer
- The points each question carries are indicated.
- Most assignments have bonus questions for extra credit, do try them out!
- **Submitting the assignment** : Download the '.ipynb' and '.pdf' files from Colab and upload them to Gradescope. Do not delete any outputs from cells before submitting. Make sure to assign pages to questions when uploading your PDF to Gradescope.
- That's about it. Happy coding!


## About the dataset
This assignment uses a dataset obtained from the JSE Data Archive that contains biological and self-reported activity traits of a sample of college students at a single university uploaded in 2013. Background Information on the dataset: http://jse.amstat.org/v21n2/froelich/eyecolorgender.txt

For this lab, the dataset has already been split into a training set `df_train` and a test set `df_test`.

In [1]:
import pandas as pd
from sklearn.neural_network import MLPClassifier
from sklearn.svm import SVC

from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.feature_extraction import DictVectorizer

from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV, ParameterGrid

import numpy as np

import warnings
warnings.filterwarnings("ignore")

In [2]:
#DOWNLOADING DATASET
!pip install gdown

import gdown

file_id_1 = "1vGiw3AZqAacW7rC3H1sKp7jbqxPTANMu"
url_1 = f"https://drive.google.com/uc?id={file_id_1}"

file_id_2 = "1MYoDg-AqqhBDuuItqfpM6r_99Zr-f9MM"
url_2 = f"https://drive.google.com/uc?id={file_id_2}"


# Download training set file
output_1 = 'lab_4_training.csv'
gdown.download(url_1, output_1, quiet=False)
print('Dataset Downloaded: lab_4_training.csv')
df_train = pd.read_csv(output_1)


# Download test set file
output_2 = 'lab_4_test.csv'
gdown.download(url_2, output_2, quiet=False)
print('Dataset Downloaded: lab_4_test.csv')
df_test=pd.read_csv(output_2)



Downloading...
From: https://drive.google.com/uc?id=1vGiw3AZqAacW7rC3H1sKp7jbqxPTANMu
To: /content/lab_4_training.csv
100%|██████████| 79.2k/79.2k [00:00<00:00, 45.7MB/s]


Dataset Downloaded: lab_4_training.csv


Downloading...
From: https://drive.google.com/uc?id=1MYoDg-AqqhBDuuItqfpM6r_99Zr-f9MM
To: /content/lab_4_test.csv
100%|██████████| 26.5k/26.5k [00:00<00:00, 38.4MB/s]

Dataset Downloaded: lab_4_test.csv


In [3]:
df_train.head()

,Unnamed: 0,gender,age,year,eyecolor,height,miles,brothers,sisters,computertime,exercise,exercisehours,musiccds,playgames,watchtv
0,577,male,20,third,hazel,72.0,180.0,0,0,5.0,No,0.0,100.0,10.0,10.0
1,677,male,19,second,hazel,72.0,120.0,1,1,16.0,Yes,9.0,70.0,3.0,5.0
2,1738,male,20,second,brown,63.0,55.0,1,2,15.0,Yes,4.5,15.0,4.0,13.0
3,1355,male,20,third,green,78.0,200.0,0,0,10.0,Yes,9.0,20.0,10.0,10.0
4,891,female,19,second,green,67.0,280.0,2,0,4.0,Yes,2.0,164.0,0.0,2.0


In [4]:
df_test.head()

,Unnamed: 0,gender,age,year,eyecolor,height,miles,brothers,sisters,computertime,exercise,exercisehours,musiccds,playgames,watchtv
0,1303,male,20,second,green,73.0,210.0,0,1,10.0,Yes,5.0,50.0,1.0,15.0
1,36,male,20,third,other,71.0,90.0,1,0,15.0,Yes,4.0,10.0,0.0,1.0
2,489,male,22,fourth,hazel,75.0,200.0,0,1,1.0,Yes,2.0,150.0,1.0,10.0
3,1415,male,19,second,brown,72.0,35.0,2,2,20.0,Yes,5.0,100.0,0.0,7.0
4,616,male,22,fourth,hazel,71.0,15.0,2,1,10.0,Yes,7.0,10.0,0.0,5.0


***
### Question 1 (1 point) ###
Calculate a baseline accuracy measure using the majority class, assuming a target variable of `gender`. The majority class is the most common value of the target variable in a particular dataset. Accuracy is calculated as (true positives + true negatives) / (all negatives and positives).

**Question 1.a**  
Find the majority class in the training set. If you always predicted this class in the training set, what would your accuracy be?

In [5]:
# YOUR CODE HERE
majority_class = df_train["gender"].mode()[0]
train_accuracy = (df_train["gender"] == majority_class).mean()

print("Majority class:", majority_class)
print("Training accuracy:", train_accuracy)

Majority class: female
Training accuracy: 0.5427852348993288


###ANSWER:_horiz

Majority class: female
Training accuracy: 0.5427852348993288


```
# This is formatted as code
```



**Question 1.b**   
If you always predicted this same class (majority from the training set) in the test set, what would your accuracy be?

In [6]:
# YOUR CODE HERE
test_accuracy = (df_test["gender"] == majority_class).mean()

print("Test accuracy:", test_accuracy)

Test accuracy: 0.5226130653266332


###ANSWER:
Test accuracy: 0.5226130653266332

***
### Question 2 (1.5 points) ###
Get started with Neural Networks.

Choose a neural-network implementation (we recommend scikit-learn's `MLPClassifier` for questions 2a-2c) and specify which implementation you use. The implementation should allow you to modify the number of hidden layers and the number of nodes in each layer.

**Note:** When possible, use the logistic (`sigmoid`) activation function and the L-BFGS optimizer (`solver='lbfgs'`).

**Question 2.a**   
Train a neural network with a single hidden layer containing `10 nodes`. Use only the `height` feature from the training set to predict `gender`. Encode `gender` as a binary class. After training, use the trained model to predict `gender` for the training observations using only `height`. What is the training accuracy?

In [8]:
# YOUR CODE HERE
# Encode gender as binary labels
label_encoder = LabelEncoder()
y_train = label_encoder.fit_transform(df_train["gender"])
y_test = label_encoder.transform(df_test["gender"])

# Select height as the only input feature
X_train = df_train[["height"]]
X_test = df_test[["height"]]

# Train the neural network
model = MLPClassifier(
    hidden_layer_sizes=(10,),
    activation="logistic",
    solver="lbfgs",
    random_state=42
)
model.fit(X_train, y_train)

# Calculate training accuracy
train_predictions = model.predict(X_train)
train_accuracy = accuracy_score(y_train, train_predictions)

print(f"Training accuracy: {train_accuracy:.4f}")
print(f"Training accuracy (%): {train_accuracy * 100:.2f}%")

Training accuracy: 0.8440
Training accuracy (%): 84.40%


###ANSWER:
Training accuracy: 0.8440
Training accuracy (%): 84.40%


**Question 2.b**  
Use the model trained in Question 2a to predict `gender` for the test set using only `height`. Do not refit the model on the test set. What is the test accuracy?

In [9]:
# YOUR CODE HERE
test_predictions = model.predict(X_test)
test_accuracy = accuracy_score(y_test, test_predictions)

print(f"Test accuracy: {test_accuracy:.4f}")
print(f"Test accuracy (%): {test_accuracy * 100:.2f}%")

Test accuracy: 0.8543
Test accuracy (%): 85.43%


###ANSWER:


Test accuracy: 0.8543
Test accuracy (%): 85.43%


**Question 2.c**   
Neural Networks often perform better when continuous input features are scaled to a smaller, consistent range. Use `MinMaxScaler` from scikit-learn to scale the `height` feature between 0 and 1. Fit the scaler using the training set, then apply that same fitted scaler to the test set. Retrain the same neural-network architecture as Questions 2.a using the scaled `height` values, and report the accuracy on both the training and test sets.

In [10]:
# YOUR CODE HERE
# Scale height using the training set's minimum and maximum
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Train a new model with the same architecture and settings
model_scaled = MLPClassifier(
    hidden_layer_sizes=(10,),
    activation="logistic",
    solver="lbfgs",
    random_state=42
)
model_scaled.fit(X_train_scaled, y_train)

# Calculate both accuracies
train_accuracy_scaled = accuracy_score(
    y_train, model_scaled.predict(X_train_scaled)
)
test_accuracy_scaled = accuracy_score(
    y_test, model_scaled.predict(X_test_scaled)
)

print(f"Scaled training accuracy: {train_accuracy_scaled * 100:.2f}%")
print(f"Scaled test accuracy: {test_accuracy_scaled * 100:.2f}%")

Scaled training accuracy: 84.40%
Scaled test accuracy: 85.43%


###ANSWER:
Scaled training accuracy: 84.40%
Scaled test accuracy: 85.43%


***

### Question 3 (1 point) ###
`MLPClassifier` requires numeric input, so the categorical features `year`, `eyecolor`, and `exercise` must first be converted into numeric features. Transform each of these variables using one-hot encoding: create one binary feature for each unique category, where the value is `1` when an observation belongs to that category and `0` otherwise. Using only these one-hot encoded features, train a Neural Network with a single hidden layer containing 10 nodes and predict `gender` for the test set. Use a maximum of 50 training iterations. What is the test accuracy?

In [11]:
# YOUR CODE HERE
# One-hot encode the three categorical features
features = ["year", "eyecolor", "exercise"]

encoder = OneHotEncoder(
    sparse_output=False,
    handle_unknown="ignore"
)

X_train_q3 = encoder.fit_transform(df_train[features])
X_test_q3 = encoder.transform(df_test[features])

# Train with one hidden layer of 10 nodes and at most 50 iterations
model_q3 = MLPClassifier(
    hidden_layer_sizes=(10,),
    activation="logistic",
    solver="lbfgs",
    max_iter=50,
    random_state=42
)
model_q3.fit(X_train_q3, y_train)

# Calculate test accuracy
test_predictions_q3 = model_q3.predict(X_test_q3)
test_accuracy_q3 = accuracy_score(y_test, test_predictions_q3)

print(f"Test accuracy: {test_accuracy_q3:.4f}")
print(f"Test accuracy (%): {test_accuracy_q3 * 100:.2f}%")

Test accuracy: 0.5528
Test accuracy (%): 55.28%


###ANSWER:

Test accuracy: 0.5528
Test accuracy (%): 55.28%


***
### Question 4 (3 points) ###
Using a NN, report the accuracy on the test set of a model trained only on `height` and the `eyecolor` features of instances in the training set.

**Question 4.a**  
What is the accuracy on the test set using `height` values and `eyecolor` as a one-hot?

In [12]:
# YOUR CODE HERE
# One-hot encode eyecolor
eye_encoder = OneHotEncoder(
    sparse_output=False,
    handle_unknown="ignore"
)

eye_train = eye_encoder.fit_transform(df_train[["eyecolor"]])
eye_test = eye_encoder.transform(df_test[["eyecolor"]])

# Combine original height with encoded eyecolor
X_train_q4a = np.column_stack((df_train["height"], eye_train))
X_test_q4a = np.column_stack((df_test["height"], eye_test))

model_q4a = MLPClassifier(
    hidden_layer_sizes=(10,),
    activation="logistic",
    solver="lbfgs",
    random_state=42
)
model_q4a.fit(X_train_q4a, y_train)

accuracy_q4a = accuracy_score(
    y_test, model_q4a.predict(X_test_q4a)
)

print(f"Test accuracy: {accuracy_q4a * 100:.2f}%")

Test accuracy: 86.18%


###ANSWER:


Test accuracy: 86.18%


**Question 4.b**  
What is the accuracy on the test set using the log of `height` values (applied to both training and testing sets) and `eyecolor` as a one-hot?

In [13]:
# YOUR CODE HERE
# Combine log-transformed height with encoded eyecolor
X_train_q4b = np.column_stack((
    np.log(df_train["height"]), eye_train
))
X_test_q4b = np.column_stack((
    np.log(df_test["height"]), eye_test
))

model_q4b = MLPClassifier(
    hidden_layer_sizes=(10,),
    activation="logistic",
    solver="lbfgs",
    random_state=42
)
model_q4b.fit(X_train_q4b, y_train)

accuracy_q4b = accuracy_score(
    y_test, model_q4b.predict(X_test_q4b)
)

print(f"Test accuracy: {accuracy_q4b * 100:.2f}%")

Test accuracy: 86.18%


###ANSWER:
Test accuracy: 86.18%

**Question 4.c**  
What is the accuracy on the test set using the Z-score of `height` values and `eyecolor` as a one-hot?

Z-score is a normalization function. It is the value of a feature minus the average value for that feature (in the training set), divided by the standard deviation of that feature (in the training set). Remember that, whenever applying a function to a feature in the training set, it also has to be applied to that same feature in the test set.

In [14]:
# YOUR CODE HERE
# Standardize height using training-set statistics
height_scaler = StandardScaler()

height_train_z = height_scaler.fit_transform(df_train[["height"]])
height_test_z = height_scaler.transform(df_test[["height"]])

# Combine standardized height with encoded eyecolor
X_train_q4c = np.column_stack((height_train_z, eye_train))
X_test_q4c = np.column_stack((height_test_z, eye_test))

model_q4c = MLPClassifier(
    hidden_layer_sizes=(10,),
    activation="logistic",
    solver="lbfgs",
    random_state=42
)
model_q4c.fit(X_train_q4c, y_train)

accuracy_q4c = accuracy_score(
    y_test, model_q4c.predict(X_test_q4c)
)

print(f"Test accuracy: {accuracy_q4c * 100:.2f}%")

Test accuracy: 86.18%


###ANSWER:


Test accuracy: 86.18%


***
### Question 5 (1.5 points) ###
Repeat question 4 for `watchtv` & `eyecolor`.

**Question 5.a**  
What is the accuracy on the test set using the original `watchtv` values (no pre-processing) and `eyecolor` as a one-hot?

In [16]:
# YOUR CODE HERE
X_train_q5a = np.column_stack((df_train["watchtv"], eye_train))
X_test_q5a = np.column_stack((df_test["watchtv"], eye_test))

model_q5a = MLPClassifier(
    hidden_layer_sizes=(10,),
    activation="logistic",
    solver="lbfgs",
    random_state=42
)
model_q5a.fit(X_train_q5a, y_train)

accuracy_q5a = accuracy_score(
    y_test, model_q5a.predict(X_test_q5a)
)

print(f"Test accuracy: {accuracy_q5a * 100:.2f}%")

Test accuracy: 50.75%


###ANSWER:
Test accuracy: 50.75%


**Question 5.b**  \
What is the accuracy on the test set using the log of `watchtv` values (applied to both training and testing sets) and `eyecolor` as a one-hot?

Comment: remove rows where `watchtv == 0` from both sets before applying `log` function in order to avoid -inf values.

In [17]:
# YOUR CODE HERE
# Identify rows to keep without changing the original datasets
train_mask = (df_train["watchtv"] != 0).to_numpy()
test_mask = (df_test["watchtv"] != 0).to_numpy()

# Combine log-transformed watchtv with matching eyecolor rows
X_train_q5b = np.column_stack((
    np.log(df_train.loc[train_mask, "watchtv"]),
    eye_train[train_mask]
))
X_test_q5b = np.column_stack((
    np.log(df_test.loc[test_mask, "watchtv"]),
    eye_test[test_mask]
))

# Keep the corresponding gender labels
y_train_q5b = y_train[train_mask]
y_test_q5b = y_test[test_mask]

model_q5b = MLPClassifier(
    hidden_layer_sizes=(10,),
    activation="logistic",
    solver="lbfgs",
    random_state=42
)
model_q5b.fit(X_train_q5b, y_train_q5b)

accuracy_q5b = accuracy_score(
    y_test_q5b, model_q5b.predict(X_test_q5b)
)

print(f"Test accuracy: {accuracy_q5b * 100:.2f}%")

Test accuracy: 50.77%


###ANSWER:
Test accuracy: 50.77%

**Question 5.c** \
What is the accuracy on the test set using the Z-score of `watchtv` values and `eyecolor` as a one-hot?

In [18]:
# YOUR CODE HERE
tv_scaler = StandardScaler()

tv_train_z = tv_scaler.fit_transform(df_train[["watchtv"]])
tv_test_z = tv_scaler.transform(df_test[["watchtv"]])

X_train_q5c = np.column_stack((tv_train_z, eye_train))
X_test_q5c = np.column_stack((tv_test_z, eye_test))

model_q5c = MLPClassifier(
    hidden_layer_sizes=(10,),
    activation="logistic",
    solver="lbfgs",
    random_state=42
)
model_q5c.fit(X_train_q5c, y_train)

accuracy_q5c = accuracy_score(
    y_test, model_q5c.predict(X_test_q5c)
)

print(f"Test accuracy: {accuracy_q5c * 100:.2f}%")

Test accuracy: 53.27%


###ANSWER:


Test accuracy: 53.27%


***
### Question 6 (2 points) ###
Combine the features from question 3, 4, and 5 (`year`, `eyecolor`, `exercise`, `height`, `watchtv`). For numeric features use the best normalization method from questions 4 and 5.

**Question 6.a**  
What was the NN accuracy on the test set using a single hidden layer with 10 nodes?

In [19]:
# YOUR CODE HERE
# One-hot encode the categorical features
categorical_features = ["year", "eyecolor", "exercise"]

encoder_q6 = OneHotEncoder(
    sparse_output=False,
    handle_unknown="ignore"
)

cat_train = encoder_q6.fit_transform(df_train[categorical_features])
cat_test = encoder_q6.transform(df_test[categorical_features])

# Z-score scale the numeric features
numeric_features = ["height", "watchtv"]

scaler_q6 = StandardScaler()

num_train = scaler_q6.fit_transform(df_train[numeric_features])
num_test = scaler_q6.transform(df_test[numeric_features])

# Combine numeric and categorical features
X_train_q6 = np.column_stack((num_train, cat_train))
X_test_q6 = np.column_stack((num_test, cat_test))

# Train a neural network with one hidden layer of 10 nodes
model_q6 = MLPClassifier(
    hidden_layer_sizes=(10,),
    activation="logistic",
    solver="lbfgs",
    random_state=42
)
model_q6.fit(X_train_q6, y_train)

# Calculate test accuracy
accuracy_q6 = accuracy_score(
    y_test, model_q6.predict(X_test_q6)
)

print(f"Test accuracy: {accuracy_q6 * 100:.2f}%")

Test accuracy: 82.41%


###ANSWER:

Test accuracy: 82.41%


***
### Question 7- Bonus (1 point) ###
Can you improve the test-set accuracy from Question 6 by at least **2 percentage points**?

Using the same input features and preprocessing as in Question 6, experiment with neural-network hyperparameters such as the number of hidden layers, the number of nodes in each hidden layer, the activation function, solver, or regularization strength. **Do not add additional input features**.

You may use a systematic tuning method such as `GridSearchCV` to compare different hyperparameter combinations. When selecting hyperparameters, use cross-validation on the training data rather than selecting a model based directly on test-set performance.

Report the test-set accuracy of your final model and write a short paragraph explaining:

- which hyperparameters you explored,
- how you selected the final model, and
- how its test accuracy compares with your result from Question 6.

If you do not reach the 2-percentage-point target, report the best result you obtained and explain your approach.




In [20]:
# YOUR CODE HERE
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import StratifiedKFold

# Same features and preprocessing as Q6.
# Putting preprocessing inside the pipeline ensures it is fitted
# separately on each training fold during cross-validation.
preprocessor = ColumnTransformer([
    ("numeric", StandardScaler(), ["height", "watchtv"]),
    ("categorical", OneHotEncoder(
        sparse_output=False,
        handle_unknown="ignore"
    ), ["year", "eyecolor", "exercise"])
])

pipeline = Pipeline([
    ("preprocessor", preprocessor),
    ("model", MLPClassifier(random_state=42, max_iter=1000))
])

# Compare 36 combinations of model settings
param_grid = {
    "model__hidden_layer_sizes": [(10,), (20,), (20, 10)],
    "model__activation": ["logistic", "relu"],
    "model__solver": ["lbfgs", "adam"],
    "model__alpha": [0.0001, 0.01, 0.1]
}

cv = StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

search = GridSearchCV(
    estimator=pipeline,
    param_grid=param_grid,
    scoring="accuracy",
    cv=cv,
    n_jobs=-1
)

features_q7 = ["height", "watchtv", "year", "eyecolor", "exercise"]

# Choose settings using only training data.
# GridSearchCV then refits the best pipeline on the full training set.
search.fit(df_train[features_q7], y_train)

# Evaluate the selected model on the test set
predictions_q7 = search.predict(df_test[features_q7])
accuracy_q7 = accuracy_score(y_test, predictions_q7)

improvement_pp = (accuracy_q7 - accuracy_q6) * 100

best_settings = {
    key.replace("model__", ""): value
    for key, value in search.best_params_.items()
}

print("Best settings:", best_settings)
print(f"Best cross-validation accuracy: {search.best_score_ * 100:.2f}%")
print(f"Q6 test accuracy: {accuracy_q6 * 100:.2f}%")
print(f"Q7 test accuracy: {accuracy_q7 * 100:.2f}%")
print(f"Change from Q6: {improvement_pp:+.2f} percentage points")
print("Reached the 2-percentage-point target:", improvement_pp >= 2)

Best settings: {'activation': 'relu', 'alpha': 0.1, 'hidden_layer_sizes': (10,), 'solver': 'adam'}
Best cross-validation accuracy: 84.65%
Q6 test accuracy: 82.41%
Q7 test accuracy: 85.18%
Change from Q6: +2.76 percentage points
Reached the 2-percentage-point target: True


###ANSWER:
I used GridSearchCV with 5-fold cross-validation on the training data to compare 36 combinations of hidden-layer sizes, activation functions, solvers, and regularization strengths. I kept the same features and preprocessing as Q6. The model with the highest mean cross-validation accuracy, 84.65%, used one hidden layer with 10 nodes, ReLU activation, the Adam solver, and alpha = 0.1. Its test accuracy was 85.18%, compared with 82.41% in Q6. Using the unrounded accuracies, this represents an improvement of 2.76 percentage points, exceeding the required 2-percentage-point target.


